File: nuxeo-core/nuxeo-core-query/src/main/java/org/nuxeo/ecm/core/query/Query.java
/*
 * (C) Copyright 2006-2009 Nuxeo SA (http://nuxeo.com/) and contributors.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 2.1 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * Contributors:
 *     Bogdan Stefanescu
 *     Florent Guillaume
 */

package org.nuxeo.ecm.core.query;

/**
 * @author Bogdan Stefanescu
 * @author Florent Guillaume
 */
public interface Query {

    /**
     * Defines general query types.
     * <p>
     * There could be Query implementations for one or another Query Type. If
     * the query factory instantiating a specific implementation of this class
     * does not support a given Query Type than a {@code
     * UnsupportedQueryTypeException} should be thrown.
     */
    enum Type {
        NXQL("NXQL"), XPATH("XPATH");

        private final String name;

        Type(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return "Query#Type: " + name;
        }
    }

    /**
     * Makes a query to the backend. No filter, permission or policy filtering
     * are done.
     *
     * @return a query result object describing the resulting documents
     * @throws QueryException
     * @see {@link FilterableQuery#execute(QueryFilter, boolean)}
     */
    QueryResult execute() throws QueryException;

    /**
     * Makes a query to the backend. No filter, permission or policy filtering
     * are done.
     * <p>
     * The total number of documents can also be retrieved, it is then stored in
     * the {@link DocumentModelList} returned by
     * {@link QueryResult#getDocumentModels}.
     *
     * @param countTotal if {@code true}, also count the total number of
     *            documents when no limit/offset is passed
     * @return a query result object describing the resulting documents
     * @throws QueryException
     * @see {@link FilterableQuery#execute(QueryFilter, boolean)}
     */
    QueryResult execute(boolean countTotal) throws QueryException;

}

File: nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/query.js
/*
 * (C) Copyright 2014 Nuxeo SA (http://nuxeo.com/) and contributors.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 2.1 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl-2.1.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * Contributors:
 *     Nelson Silva <nelson.silva@inevo.pt>
 */
import {Request} from './request';

const PATH = '/query';

/**
 * REST Query
 */
class Query extends Request {

  constructor(connection) {
    super(connection, PATH);
  }

  set nxql(q) {
    this.path = `${PATH}/NXQL/`;
    this.params.query = q;
  }

  set cmisql(q) {
    this.path = `${PATH}/CMIS/`;
    this.params.query = q;
  }

  set queryParams(p) {
    this.params.queryParams = p;
  }

  set pageProvider(name) {
    this.path = `${PATH}/pageprovider/`;
    this.params.name = name;
  }

  set page(p) {
    this.params.page = p;
  }

  set pageSize(s) {
    this.params.pageSize = s;
  }

  run() {
    return this.execute();
  }
}

export {Query};
